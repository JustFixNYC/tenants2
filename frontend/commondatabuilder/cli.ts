import { argvHasOption } from "../querybuilder/util";
import ourConfig from "../../common-data/config";
import { createCommonDataFiles } from "./commondatabuilder";

const VERSION = "0.1.0";

export function run() {
  if (argvHasOption("-h", "--help")) {
    console.log(`usage: ${process.argv[1]} [OPTIONS]\n`);
    console.log(`options:\n`);
    console.log("  -h / --help     Show this help");
    console.log("  -v / --version  Show the version number");
    process.exit(0);
  }

  if (argvHasOption("-v", "--version")) {
    console.log(`commondatabuilder version ${VERSION}`);
    process.exit(0);
  }

  createCommonDataFiles(ourConfig);
}
