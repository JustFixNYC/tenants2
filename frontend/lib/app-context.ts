import React from 'react';

import { AllSessionInfo } from './queries/AllSessionInfo';

export interface AppServerInfo {
  /**
   * The URL of the server's static files, e.g. "/static/".
   */
  staticURL: string;

  /**
   * The URL to generated webpack bundles for lazy-loading, e.g. "/static/frontend/".
   */
  webpackPublicPathURL: string;

  /**
   * The URL of the server's Django admin, e.g. "/admin/".
   */
  adminIndexURL: string;

  /** The batch GraphQL endpoint; required if a GraphQL client is not provided. */
  batchGraphQLURL: string;

  /**
   * Whether the site is in development mode (corresponds to settings.DEBUG in
   * the Django app).
   */
  debug: boolean;
}

export interface AppContextType {
  server: AppServerInfo;
  session: AllSessionInfo;
}

class UnimplementedError extends Error {}

export const defaultContext: AppContextType = {
  get server(): AppServerInfo {
    throw new UnimplementedError();
  },
  get session(): AllSessionInfo {
    throw new UnimplementedError();
  }
};

export const AppContext = React.createContext<AppContextType>(defaultContext);
