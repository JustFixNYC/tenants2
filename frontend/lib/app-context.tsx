import React from 'react';

import { AllSessionInfo } from './queries/AllSessionInfo';
import { GraphQLFetch } from './graphql-client';
import { Omit } from './util';

/** Metadata about forms submitted via legacy POST. */
export interface AppLegacyFormSubmission {
  /** The original form input. */
  input: any;

  /** The result of the GraphQL mutation for the form. */
  result: any;
}

/** Details about the server that don't change through the app's lifetime. */
export interface AppServerInfo {
  /**
   * The URL of the server's static files, e.g. "/static/".
   */
  staticURL: string;

  /**
   * The URL to generated webpack bundles for lazy-loading, e.g. "/static/frontend/".
   */
  webpackPublicPathURL: string;

  /**
   * The URL of the server's Django admin, e.g. "/admin/".
   */
  adminIndexURL: string;

  /** The batch GraphQL endpoint; required if a GraphQL client is not provided. */
  batchGraphQLURL: string;

  /** The letter of complaint URL (HTML format). */
  locHtmlURL: string;

  /** The letter of complaint URL (PDF format). */
  locPdfURL: string;

  /**
   * Whether the site is in development mode (corresponds to settings.DEBUG in
   * the Django app).
   */
  debug: boolean;
}

/**
 * Basic information about the app that components
 * should have relatively easy access to.
 */
export interface AppContextType {
  /**
   * Information about the server that stays constant through the app's
   * lifetime.
   */
  server: AppServerInfo;

  /**
   * Information about the current user that may change if they
   * log in/out, etc.
   */
  session: AllSessionInfo;

  /**
   * A reference to the app's GraphQL interface, for network requets.
   */
  fetch: GraphQLFetch;

  /**
   * Currently, we often update the app's state by having network
   * APIs return a new session state. This makes it easy for
   * components to just pass the session data back to the app.
   */
  updateSession: (session: Partial<AllSessionInfo>) => void;

  /** If a form was submitted via a non-JS browser, data will be here. */
  legacyFormSubmission?: AppLegacyFormSubmission;
}

/* istanbul ignore next: this will never be executed in practice. */
class UnimplementedError extends Error {
  constructor() {
    super("This is unimplemented!");
  }
}

/* istanbul ignore next: this will never be executed in practice. */
/**
 * The default AppContext will raise an exception when any of its
 * properties are accessed; because this information is very
 * important to the user experience, we really need it to be
 * provided by the app!
 * 
 * However, we're also exporting the symbol, so test suites
 * can use Object.defineProperty() to override the properties
 * and provide defaults for testing. This will ensure that
 * tests don't need to wrap everything in an AppContext.Provider.
 */
export const defaultContext: AppContextType = {
  get server(): AppServerInfo {
    throw new UnimplementedError();
  },
  get session(): AllSessionInfo {
    throw new UnimplementedError();
  },
  fetch(query: string, variables?: any): Promise<any> {
    throw new UnimplementedError();
  },
  updateSession(session: Partial<AllSessionInfo>) {
    throw new UnimplementedError();
  }
};

/**
 * A React Context that provides basic information about
 * the app that we don't want to have to pass down through
 * our whole component heirarchy.
 * 
 * For more details, see:
 * 
 *   https://reactjs.org/docs/context.html
 */
export const AppContext = React.createContext<AppContextType>(defaultContext);

/**
 * Higher-order component (HOC) factory function to wrap existing
 * components in a context consumer.
 */
export function withAppContext<P extends AppContextType>(Component: React.ComponentType<P>): React.ComponentType<Omit<P, keyof AppContextType>> {
  return function(props: Omit<P, keyof AppContextType>) {
    return (
      <AppContext.Consumer>
        {(context) => <Component {...props} {...context} />}
      </AppContext.Consumer>
    );
  }
}
