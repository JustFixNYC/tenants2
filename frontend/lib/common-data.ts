/**
 * This module provides types and functionality to interact with common data used
 * by both the front-end and back-end.
 */

 /**
  * This data type is used by Django to define a set of choices for a form or
  * model field. The first string is the machine-readable value of the field that
  * is used in data serialization, while the second string is the human-readable
  * label.
  * 
  * For more details, see:
  * 
  *   https://docs.djangoproject.com/en/2.1/ref/models/fields/#choices
  */
export type DjangoChoice = [string, string];

export type DjangoChoices = DjangoChoice[];


/**
 * An enhancement of DjangoChoice that makes it more flexible for use in
 * React-based user interfaces.
 */
export type ReactDjangoChoice = [string, string|JSX.Element];

export type ReactDjangoChoices = ReactDjangoChoice[];


/**
 * Retrieve the human-readable label for a choice, given its machine-readable value.
 * 
 * Throw an exception if the choice is invalid.
 */
export function getDjangoChoiceLabel(choices: DjangoChoices, value: string): string {
  const result = safeGetDjangoChoiceLabel(choices, value);
  if (result === null) {
    throw new Error(`Unable to find label for value ${value}`);
  }
  return result;
}

/**
 * Filter out the given values from either the given list of choices, or anything
 * that matches the given regular expression.
 */
export function filterDjangoChoices(choices: DjangoChoices, values: string[]|RegExp): DjangoChoices {
  if (Array.isArray(values)) {
    if (process.env.NODE_ENV !== 'production') {
      validateDjangoChoices(choices, values);
    }
    return choices.filter(([value, _]) => !values.includes(value));
  } else {
    return choices.filter(([value, _]) => !values.test(value));
  }
}

/**
 * Retrieve the human-readable label for a choice, given its machine-readable value.
 * 
 * Return null if the choice is invalid.
 */
export function safeGetDjangoChoiceLabel(choices: DjangoChoices, value: string): string|null {
  for (let [v, label] of choices) {
    if (v === value) return label;
  }
  return null;
}

/**
 * Validate that the given values are valid choices.
 * 
 * This is intended to be used in tests. It should be removed from
 * production bundles via tree-shaking.
 */
export function validateDjangoChoices(choices: DjangoChoices, values: string[]) {
  values.forEach(value => {
    getDjangoChoiceLabel(choices, value);
  });
}

/**
 * Convert an all-caps value, like 'FOO_BAR', to an
 * all-lowercase slug-friendly value, like 'foo-bar'.
 */
export function allCapsToSlug(value: string): string {
  return value.toLowerCase().replace(/_/g, '-');
}

/**
 * Convert an all-lowercase slug-friendly value,
 * like 'foo-bar', to an all-caps value, like
 * 'FOO_BAR'.
 */
export function slugToAllCaps(value: string): string {
  return value.toUpperCase().replace(/-/g, '_');
}

type StringMapping<T extends string> = {
  [k in T]: string
};

export function toDjangoChoices<T extends string>(choices: T[], labels: StringMapping<T>): [string, string][] {
  return choices.map(choice => [choice, labels[choice]] as [string, string]);
}
