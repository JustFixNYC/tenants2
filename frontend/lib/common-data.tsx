export type DjangoChoice = [string, string];

export type DjangoChoices = DjangoChoice[];
