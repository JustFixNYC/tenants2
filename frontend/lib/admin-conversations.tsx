import React, { useEffect, useContext, useState, useMemo } from 'react';
import { Switch, Route } from "react-router-dom";
import Routes from "./routes";
import { AppContext } from './app-context';
import { AdminConversations, AdminConversationsVariables } from './queries/AdminConversations';
import { QueryLoaderQuery } from './query-loader-prefetcher';
import { AdminConversationVariables, AdminConversation } from './queries/AdminConversation';

function useQuery<Input, Output>(query: QueryLoaderQuery<Input, Output>, input: Input|null): Output|null {
  const [value, setValue] = useState<Output|null>(null);
  const appCtx = useContext(AppContext);
  const { fetch } = appCtx;

  useEffect(() => {
    let isMounted = true;
    if (input !== null) {
      const result = query.fetch(fetch, input);
      result.then(v => isMounted && setValue(v));
    }
    return () => { isMounted = false; };
  }, [fetch, input]);

  return value;
}

const AdminConversationsPage: React.FC<{}> = (props) => {
  const conversationsInput = useMemo<AdminConversationsVariables>(() => ({
    query: '',
    page: 1,
  }), []);
  const conversations = useQuery(AdminConversations, conversationsInput);
  const [selectedPhoneNumber, setSelectedPhoneNumber] = useState<string|null>(null);
  const conversationInput = useMemo<AdminConversationVariables|null>(() => selectedPhoneNumber ? {
    phoneNumber: selectedPhoneNumber,
    page: 1,
  } : null, [selectedPhoneNumber]);
  const conversation = useQuery(AdminConversation, conversationInput);

  return <div className="jf-admin-conversations-wrapper">
    <div className="jf-conversation-sidebar">
      {conversations?.output?.map(conv => {
        return <div key={conv.userPhoneNumber}
                    onClick={() => setSelectedPhoneNumber(conv.userPhoneNumber)}>
          <div>{conv.userPhoneNumber}</div>
          <div>{conv.body}</div>
        </div>
      }) || <p>Loading...</p>}
    </div>
    <div className="jf-current-conversation">
      {conversation?.output?.map(msg => {
        return <div key={msg.sid} className={msg.isFromUs ? 'jf-from-us' : 'jf-to-us'}>
          <div title={`This message was sent on ${msg.dateSent}.`}>
            {msg.body}
          </div>
        </div>
      })}
    </div>
  </div>;
};

export default function AdminConversationsRoutes(): JSX.Element {
  return (
    <Switch>
      <Route path={Routes.adminConversations} exact component={AdminConversationsPage} />
    </Switch>
  );
}
