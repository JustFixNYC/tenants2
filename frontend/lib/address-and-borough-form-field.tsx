import React from 'react';
import { LabelRenderer, BaseFormFieldProps, TextualFormField, RadiosFormField } from './form-fields';
import { toDjangoChoices } from './common-data';
import { BoroughChoices, getBoroughChoiceLabels } from '../../common-data/borough-choices';
import { ProgressiveEnhancementContext, ProgressiveEnhancement } from './progressive-enhancement';
import { safeGetBoroughChoice } from './pages/onboarding-step-1';
import { GeoAutocomplete } from './geo-autocomplete';

const DEFAULT_ADDRESS_LABEL = "Address";

type AddressAndBoroughFieldProps = {
  disableProgressiveEnhancement?: boolean;
  addressLabel?: string,
  onChange?: () => void;
  renderAddressLabel?: LabelRenderer,
  addressProps: BaseFormFieldProps<string>,
  boroughProps: BaseFormFieldProps<string>
};

export class AddressAndBoroughField extends React.Component<AddressAndBoroughFieldProps> {
  renderBaselineAddressFields(): JSX.Element {
    return (
      <React.Fragment>
        <TextualFormField
          label={this.props.addressLabel || DEFAULT_ADDRESS_LABEL}
          renderLabel={this.props.renderAddressLabel}
          {...this.props.addressProps}
        />
        <RadiosFormField
          label="What is your borough?"
          {...this.props.boroughProps}
          choices={toDjangoChoices(BoroughChoices, getBoroughChoiceLabels())}
        />
      </React.Fragment>
    );
  }

  renderEnhancedAddressField(pe: ProgressiveEnhancementContext) {
    const { addressProps, boroughProps } = this.props;
    let initialValue = addressProps.value
      ? { address: addressProps.value,
          borough: safeGetBoroughChoice(boroughProps.value) }
      : undefined;

    if (boroughProps.errors && !addressProps.errors) {
      return this.renderBaselineAddressFields();
    }

    return <GeoAutocomplete
      label={this.props.addressLabel || DEFAULT_ADDRESS_LABEL}
      renderLabel={this.props.renderAddressLabel}
      initialValue={initialValue}
      onChange={selection => {
        this.props.onChange && this.props.onChange();
        addressProps.onChange(selection.address);
        boroughProps.onChange(selection.borough || '');
      }}
      onNetworkError={pe.fallbackToBaseline}
      errors={addressProps.errors}
    />;
  }

  render() {
    return (
      <ProgressiveEnhancement
        disabled={this.props.disableProgressiveEnhancement}
        renderBaseline={() => this.renderBaselineAddressFields()}
        renderEnhanced={(pe) => this.renderEnhancedAddressField(pe)} />
    );
  }
}
