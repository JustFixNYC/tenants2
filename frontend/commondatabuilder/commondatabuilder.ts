import * as fs from 'fs';
import * as path from 'path';

import { DjangoChoices } from "../lib/common-data";

type CreateOptions = {
  exportLabels: boolean
};

const defaultOptions: CreateOptions = { exportLabels: true };

export type DjangoChoicesTypescriptConfig = {
  rootDir: string,
  files: DjangoChoicesTypescriptFileConfig[]
};

export type DjangoChoicesTypescriptFileConfig = {
  jsonFilename: string,
  typeName: string,
  exportLabels: boolean,
  filterOut?: RegExp|string[],
};

/**
 * Retrieve the human-readable label for a choice, given its machine-readable value.
 * 
 * Return null if the choice is invalid.
 */
export function safeGetDjangoChoiceLabel(choices: DjangoChoices, value: string): string|null {
  for (let [v, label] of choices) {
    if (v === value) return label;
  }
  return null;
}

/**
 * Validate that the given values are valid choices.
 * 
 * This is intended to be used in tests. It should be removed from
 * production bundles via tree-shaking.
 */
export function validateDjangoChoices(choices: DjangoChoices, values: string[]) {
  values.forEach(value => {
    getDjangoChoiceLabel(choices, value);
  });
}

/**
 * Filter out the given values from either the given list of choices, or anything
 * that matches the given regular expression.
 */
export function filterDjangoChoices(choices: DjangoChoices, values: string[]|RegExp): DjangoChoices {
  if (Array.isArray(values)) {
    validateDjangoChoices(choices, values);
    return choices.filter(([value, _]) => !values.includes(value));
  } else {
    return choices.filter(([value, _]) => !values.test(value));
  }
}

/**
 * Retrieve the human-readable label for a choice, given its machine-readable value.
 * 
 * Throw an exception if the choice is invalid.
 */
export function getDjangoChoiceLabel(choices: DjangoChoices, value: string): string {
  const result = safeGetDjangoChoiceLabel(choices, value);
  if (result === null) {
    throw new Error(`Unable to find label for value ${value}`);
  }
  return result;
}

function replaceExt(filename: string, ext: string) {
  // https://stackoverflow.com/a/5953384
  return filename.substr(0, filename.lastIndexOf(".")) + ext;
}

export function createDjangoChoicesTypescriptFiles(
  config: DjangoChoicesTypescriptConfig,
  dryRun: boolean = false
) {
  config.files.forEach(fileConfig => {
    const infile = path.join(config.rootDir, fileConfig.jsonFilename);
    let choices = JSON.parse(fs.readFileSync(infile, {
      encoding: 'utf-8'
    })) as DjangoChoices;
    if (fileConfig.filterOut) {
      choices = filterDjangoChoices(choices, fileConfig.filterOut);
    }
    const ts = createDjangoChoicesTypescript(choices, fileConfig.typeName, {
      exportLabels: fileConfig.exportLabels
    });
    const outfilename = replaceExt(fileConfig.jsonFilename, '.ts');
    const outfile = path.join(config.rootDir, outfilename);
    if (!dryRun) {
      console.log(`Writing ${outfilename}.`);
      fs.writeFileSync(outfile, ts, { encoding: 'utf-8' });
    }
  });
}

/**
 * Return a list of TypeScript lines that create a function
 * which returns a mapping from choice values to their labels.
 * 
 * Note that this is a function rather than a constant because
 * we want to be able to add code that localizes the
 * labels at runtime if needed.
 */
function createLabelExporter(choices: DjangoChoices, name: string): string[] {
  const lines = [];
  lines.push(
    `export type ${name}Labels = {`,
    `  [k in ${name}]: string;`,
    `};\n`
  );
  lines.push(
    `export function get${name}Labels(): ${name}Labels {`,
    `  return {`
  );
  for (let [name, label] of choices) {
    lines.push(`    ${name}: ${JSON.stringify(label)},`);
  }
  lines.push(
    '  };',
    '}\n'
  );
  return lines;
}

/**
 * Return a list of TypeScript lines that define a type
 * and some helper functions for the given Django choices. 
 */
export function createDjangoChoicesTypescript(
  choices: DjangoChoices,
  name: string,
  options: Partial<CreateOptions> = {}
): string {
  const { exportLabels } = Object.assign({}, defaultOptions, options);
  const lines = [
    `// This file was auto-generated by commondatabuilder.`,
    `// Please don't edit it.\n`,
  ];
  const choiceKeys = choices.map(choice => choice[0]);
  const quotedChoiceKeys = choiceKeys.map(key => JSON.stringify(key));
  lines.push(`export type ${name} = ${quotedChoiceKeys.join('|')};\n`);
  lines.push(`export const ${name}s: ${name}[] = ${JSON.stringify(choiceKeys, null, 2)};\n`);
  lines.push(`const ${name}Set: Set<String> = new Set(${name}s);\n`);
  lines.push(
    `export function is${name}(choice: string): choice is ${name} {`,
    `  return ${name}Set.has(choice);`,
    `}\n`
  );
  if (exportLabels) {
    lines.push(...createLabelExporter(choices, name));
  }
  return lines.join('\n');
}
