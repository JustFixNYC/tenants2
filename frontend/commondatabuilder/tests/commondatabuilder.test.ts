import { createDjangoChoicesTypescript, createDjangoChoicesTypescriptFiles, getDjangoChoiceLabel, validateDjangoChoices, filterDjangoChoices } from "../commondatabuilder";

import ourConfig from "../../../common-data/config";
import { DjangoChoices } from "../../lib/common-data";

describe('commondatabuilder', () => {
  it('creates django choice typescript', () => {
    const ts = createDjangoChoicesTypescript([['BROOKLYN', 'Brooklyn']], 'BoroughChoice');
    expect(ts).toMatchSnapshot();
  });

  it('only exports labels if configured to', () => {
    expect(createDjangoChoicesTypescript([], 'Foo')).toMatch(/getFooLabels/);
    expect(createDjangoChoicesTypescript([], 'Foo', {
      exportLabels: false
    })).not.toMatch(/getFooLabels/);
  });

  it('works with our common data configuration', () => {
    createDjangoChoicesTypescriptFiles(ourConfig, true);
  });
});

test('getDjangoChoiceLabel() works', () => {
  expect(getDjangoChoiceLabel([['BLAH', 'boop']], 'BLAH')).toBe('boop');
  expect(() => getDjangoChoiceLabel([['BLAH', 'boop']], 'NOPE'))
    .toThrow('Unable to find label for value NOPE');
});

test("validateDjangoChoices() works", () => {
  expect(validateDjangoChoices([['BLAH', 'boop']], ['BLAH'])).toBeUndefined();
  expect(() => validateDjangoChoices([['BLAH', 'boop']], ['NOPE']))
    .toThrow('Unable to find label for value NOPE');
});

test("filterDjangoChoices() works", () => {
  const choices: DjangoChoices = [["FOO", "Foo"], ["BAR", "Bar"]];
  expect(filterDjangoChoices(choices, ["FOO"])).toEqual([["BAR", "Bar"]]);
  expect(filterDjangoChoices(choices, /^BA/)).toEqual([["FOO", "Foo"]]);
});
