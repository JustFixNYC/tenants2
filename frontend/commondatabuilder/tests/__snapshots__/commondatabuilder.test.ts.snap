// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commondatabuilder creates django choice typescript 1`] = `
"// This file was auto-generated by commondatabuilder.
// Please don't edit it.

export type BoroughChoice = \\"BROOKLYN\\";

export const BoroughChoices: BoroughChoice[] = [
  \\"BROOKLYN\\"
];

const BoroughChoiceSet: Set<String> = new Set(BoroughChoices);

export function isBoroughChoice(choice: string): choice is BoroughChoice {
  return BoroughChoiceSet.has(choice);
}

export type BoroughChoiceLabels = {
  [k in BoroughChoice]: string;
};

export function getBoroughChoiceLabels(): BoroughChoiceLabels {
  return {
    BROOKLYN: \\"Brooklyn\\",
  };
}
"
`;

exports[`commondatabuilder only internationalizes labels if configured to 1`] = `
"// This file was auto-generated by commondatabuilder.
// Please don't edit it.

import { t } from \\"@lingui/macro\\";
import { li18n } from '../frontend/lib/i18n-lingui';

export type Foo = \\"THINGY\\";

export const Foos: Foo[] = [
  \\"THINGY\\"
];

const FooSet: Set<String> = new Set(Foos);

export function isFoo(choice: string): choice is Foo {
  return FooSet.has(choice);
}

export type FooLabels = {
  [k in Foo]: string;
};

export function getFooLabels(): FooLabels {
  return {
    THINGY: li18n._(t\`I am thingy\`),
  };
}
"
`;
